### PR TITLE
Fix #1024 : Exploration back flow

### LINKED ISSUE #1024: Exploration back flow is incorrect.
Exploration can be opened from places:
1. Recently Played Stories
2. Topic-Lessons-Chapter Name
3. StoryActivity -> Chapter Item

All these different flows open exploration, but on click on back button, it should always open `TopicActivity` only. This has been implemented but now there seems inconsistency in that.

For example:
1. Go to Ratios topic - Lessons tab - expand an item and open exploration. Leave the exploration. You will notice two issues:
- (a.) Info tab is visible but ideally Lessons tab should be visible.
- (b.) Click on back again, you will see TopicActivity with Lessons Tab and the story list is duplicated.

2. Go to Ratios topic - Lessons tab - Click on Item to open `StoryActivity` (chapters with images). Open exploration. Leave exploration. Now you will notice `TopicActivity` appears which is correct. Click on back on this screen and you will see `StoryActivity` and click back again and you will see `TopicActivity`. This should not happen.

### PR DESCRIPTION
**Pull request by Sarthak2601**
## Explanation

- Fix #1024 

- The TopicFragmentPresenter only set the Lession tab as the starting tab of the viewPager if both topicID and storyID returned .isNotEmpty() as true. But since story id wasn't getting returned by the stopExploration() so it set the default tab (Info) for viewPager.

- I also added the onBackPressed() to the Topic Activity so that it doesn't create any further bugs like unnecessary re-opening of topics(as opened by user) on back pressed. 

- I would like to know if I should also implement an AlertDialog on the back press of the Topic Activity ? 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by rt4914 — CHANGES_REQUESTED** [review]
Following flows are incorrect:

1. Click on Topic on homescreen should open `Info` page and not `PromotedStory`
Screenshot_1588566559


2. Open Lessons tab -> Click on an item to open `StoryActivity` which contains the chapter list with images. Click on chapter - exploration starts. Leave exploration. In this case `StoryActivity` should appear instead of `TopicActivity` with Lessons tab.
story_activity


All other flows are correct but these two flows are still incorrect. Maybe you can introduce two different enum to decide these screens:
1. For deciding which activity to open in back flow
```
enum class ParentActivityForExploration(val value: Int) {
  STORY_ACTIVITY(0),
  TOPIC_ACTIVITY(1);
}

```

2. For deciding which tab to open in TopicActivity in back flow
```
enum class TopicActivityStartingTab(val value: Int) {
  INFO(0),
  LESSONS(1);
}
```

You can pass these enums to `ExplorationActivity` and use this information to control which screen to show on back press.

**Comment by Sarthak2601** [comment]
> Following flows are incorrect:
> 
> 1. Click on Topic on homescreen should open `Info` page and not `PromotedStory`
>    Screenshot_1588566559
> 2. Open Lessons tab -> Click on an item to open `StoryActivity` which contains the chapter list with images. Click on chapter - exploration starts. Leave exploration. In this case `StoryActivity` should appear instead of `TopicActivity` with Lessons tab.
>    story_activity
> 
> All other flows are correct but these two flows are still incorrect. Maybe you can introduce two different enum to decide these screens:
> 
> 1. For deciding which activity to open in back flow
> 
> ```
> enum class ParentActivityForExploration(val value: Int) {
>   STORY_ACTIVITY(0),
>   TOPIC_ACTIVITY(1);
> }
> ```
> 
> 1. For deciding which tab to open in TopicActivity in back flow
> 
> ```
> enum class TopicActivityStartingTab(val value: Int) {
>   INFO(0),
>   LESSONS(1);
> }
> ```
> 
> You can pass these enums to `ExplorationActivity` and use this information to control which screen to show on back press.

Sorry I took a pass there, will wrap it up soon !

**Comment by Sarthak2601** [comment]
@rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well. 
PTAL, thanks :)

**Comment by rt4914** [comment]
> @rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well.
> PTAL, thanks :)

@Sarthak2601
Application is crashing for this flow:
`Home -> View All -> RecentlyPlayedActivity -> Select any item` It will crash the application as the "exploration_id" is null.

**Comment by rt4914** [comment]
> > @rt4914 Both the flows are now corrected. I also added onBackPressed() to the StoryActivity as there was some error in that as well.
> > PTAL, thanks :)
> 
> Application is crashing for this flow:
> `Home -> View All -> RecentlyPlayedActivity -> Select any item` It will crash the application as the "exploration_id" is null.

**Comment by Sarthak2601** [comment]
@rt4914 Handled the crash and corrected the back flow of the Exploration activity opened via RecentlyPlayed stories to open the info tab as well.

File: app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt [hunk 1]
```diff
@@ -98,7 +98,7 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
         )
         else -> {
           logger.d("RecentlyPlayedFragment", "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId,null)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Nit: whenever we pass null or an integer or a boolean value we write inline comment to make it more readable code.
`routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId, /* backflowId= */ null)`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt [hunk 2]
```diff
@@ -38,7 +39,8 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     topicId = intent.getStringExtra(EXPLORATION_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)
     storyId = intent.getStringExtra(EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY)
     explorationId = intent.getStringExtra(EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY)
-    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId)
+    backflowId = intent.getStringExtra(EXPLORATION_ACTIVITY_BACKFLOW_ID)
+    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId, backflowId, this)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: `Context` should be passed as first parameter instead of last.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Great implementation. Your entire code approach is correct but instead of using string constants you should use `enum` for backflow ids.

You can check `AdminAuthEnum` and how it is being used in similar way.

Also, instead of `backflowId` you should name it as `backflowScreen` because the value is a screen name and not an id.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt (hunk 2 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Done :)

File: app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 1 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Okay, done :)

**Comment by Sarthak2601** [comment]
> Great implementation. Your entire code approach is correct but instead of using string constants you should use `enum` for backflow ids.
> 
> You can check `AdminAuthEnum` and how it is being used in similar way.
> 
> Also, instead of `backflowId` you should name it as `backflowScreen` because the value is a screen name and not an id.

@rt4914 Done with all the changes, PTAL :)

File: app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt [hunk 3]
```diff
@@ -98,7 +98,7 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
         )
         else -> {
           logger.d("RecentlyPlayedFragment", "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+          routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */null)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Nit: Spacing issue there should be space before `null`
`/* backflowScreen = */ null`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt [hunk 4]
```diff
@@ -38,7 +39,8 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     topicId = intent.getStringExtra(EXPLORATION_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)
     storyId = intent.getStringExtra(EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY)
     explorationId = intent.getStringExtra(EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY)
-    explorationActivityPresenter.handleOnCreate(internalProfileId, topicId, storyId, explorationId)
+    backflowScreen = intent.getIntExtra(EXPLORATION_ACTIVITY_BACKFLOW_SCREEN,-1)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Spacing issue: space after `,`
`intent.getIntExtra(EXPLORATION_ACTIVITY_BACKFLOW_SCREEN ,-1)`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt [hunk 5]
```diff
@@ -48,18 +50,22 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     internal const val EXPLORATION_ACTIVITY_TOPIC_ID_ARGUMENT_KEY = "ExplorationActivity.topic_id"
     internal const val EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY = "ExplorationActivity.story_id"
     internal const val EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY = "ExplorationActivity.exploration_id"
+    internal const val EXPLORATION_ACTIVITY_BACKFLOW_SCREEN = "ExplorationActivity.backflow_screen"
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Rename this to `EXPLORATION_ACTIVITY_BACKFLOW_SCREEN_KEY`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt [hunk 6]
```diff
@@ -131,6 +145,14 @@ class ExplorationActivityPresenter @Inject constructor(
     return ephemeralStateResult.getOrDefault(Exploration.getDefaultInstance())
   }
 
+  private fun backPressActivitySelector(backflowScreen: Int?){
+    when(backflowScreen){
+      ParentActivityForExploration.BACKFLOW_SCREEN_STORY.value -> activity.startActivity(StoryActivity.createStoryActivityIntent(context,internalProfileId,topicId,storyId))
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt]: Spacing issue: `createStoryActivityIntent(context, internalProfileId, topicId, storyId)`

Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt]: Ditto for below two lines, there should be space after `,`

File: app/src/main/java/org/oppia/app/story/StoryFragment.kt [hunk 7]
```diff
@@ -48,8 +48,8 @@ class StoryFragment : InjectableFragment(), ExplorationSelectionListener {
     return storyFragmentPresenter.handleCreateView(inflater, container, internalProfileId, topicId, storyId)
   }
 
-  override fun selectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
-    storyFragmentPresenter.handleSelectExploration(internalProfileId, topicId, storyId, explorationId)
+  override fun selectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowId: Int?) {
+    storyFragmentPresenter.handleSelectExploration(internalProfileId, topicId, storyId, explorationId, backflowId)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/story/StoryFragment.kt]: Rename `backflowId ` to `backflowScreen` here and everywhere else.

File: app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt [hunk 8]
```diff
@@ -68,8 +68,8 @@ class StoryFragmentPresenter @Inject constructor(
     return binding.root
   }
 
-  fun handleSelectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
-    routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId)
+  fun handleSelectExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowId: Int?) {
+    routeToExplorationListener.routeToExploration(internalProfileId, topicId, storyId, explorationId,backflowId)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt]: Rename `backflowId ` to `backflowScreen` here and everywhere else.

Comment (rt4914) [reply app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt]: Also check the space issue before `backflowId `

File: app/src/main/java/org/oppia/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt [hunk 9]
```diff
@@ -39,7 +40,7 @@ class StoryChapterSummaryViewModel(
         result.isFailure() -> logger.e(STORY_VIEWER_TAG, "Failed to load exploration", result.getErrorOrNull()!!)
         else -> {
           logger.d(STORY_VIEWER_TAG, "Successfully loaded exploration: $explorationId")
-          explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId)
+          explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */1 )
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt]: Spacing issue:
`explorationSelectionListener.selectExploration(internalProfileId, topicId, storyId, explorationId, /* backflowScreen = */ 1)`

File: app/src/main/java/org/oppia/app/testing/ExplorationTestActivity.kt [hunk 10]
```diff
@@ -17,14 +17,15 @@ class ExplorationTestActivity : InjectableAppCompatActivity(), RouteToExploratio
     explorationTestActivityPresenter.handleOnCreate()
   }
 
-  override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String) {
+  override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowScreen: Int? ) {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/testing/ExplorationTestActivity.kt]: Spacing issue: Remove space before `)`
`override fun routeToExploration(internalProfileId: Int, topicId: String, storyId: String, explorationId: String, backflowScreen: Int?)`

File: app/src/main/java/org/oppia/app/testing/ExplorationTestActivityPresenter.kt [hunk 11]
```diff
@@ -48,7 +48,7 @@ class ExplorationTestActivityPresenter @Inject constructor(
         result.isFailure() -> logger.e(TAG_EXPLORATION_TEST_ACTIVITY, "Failed to load exploration", result.getErrorOrNull()!!)
         else -> {
           logger.d(TAG_EXPLORATION_TEST_ACTIVITY, "Successfully loaded exploration")
-          routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID)
+          routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID,/* backflowScreen= */null)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/testing/ExplorationTestActivityPresenter.kt]: Spacing issue: `routeToExplorationListener.routeToExploration(INTERNAL_PROFILE_ID, TOPIC_ID, STORY_ID, EXPLORATION_ID, /* backflowScreen= */ null)`

File: app/src/main/java/org/oppia/app/topic/TopicActivity.kt [hunk 12]
```diff
@@ -48,18 +49,23 @@ class TopicActivity : InjectableAppCompatActivity(), RouteToQuestionPlayerListen
...
         storyId,
-        explorationId
+        explorationId,
+        backflowScreen
       )
     )
   }
 
+  override fun onBackPressed() {
+    startActivity(HomeActivity.createHomeActivity(this,internalProfileId))
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/TopicActivity.kt]: Spacing issue: `startActivity(HomeActivity.createHomeActivity(this, internalProfileId))`

File: app/src/main/java/org/oppia/app/topic/lessons/TopicLessonsFragmentPresenter.kt [hunk 13]
```diff
@@ -118,10 +119,10 @@ class TopicLessonsFragmentPresenter @Inject constructor(
   }
 
   override fun selectChapterSummary(storyId: String, chapterSummary: ChapterSummary) {
-    playExploration(internalProfileId, topicId, storyId, chapterSummary.explorationId)
+    playExploration(internalProfileId, topicId, storyId, chapterSummary.explorationId, /*backflowScreen =*/0)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/lessons/TopicLessonsFragmentPresenter.kt]: Spacing issue : `/*backflowScreen =*/ 0`

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt [hunk 14]
```diff
@@ -285,7 +285,7 @@ class ExplorationActivityTest {
 
   private fun createExplorationActivityIntent(internalProfileId: Int, topicId: String, storyId: String, explorationId: String): Intent {
     return ExplorationActivity.createExplorationActivityIntent(
-      ApplicationProvider.getApplicationContext(), internalProfileId, topicId, storyId, explorationId
+      ApplicationProvider.getApplicationContext(), internalProfileId, topicId, storyId, explorationId,null
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Spacing and comment issue: `explorationId, /* backflowScreen= */ null`

**Review by rt4914 — CHANGES_REQUESTED** [review]
All changes are based on spacing issue, naming issue only. Update these and also self-review this PR once this is done, so that it becomes easy to merge.

File: app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 3 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt (hunk 4 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Done

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt (hunk 5 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Done

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt (hunk 6 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/story/StoryFragment.kt (hunk 7 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/story/StoryFragment.kt]: Done

File: app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt (hunk 8 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/story/StoryFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt (hunk 9 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt]: Done

File: app/src/main/java/org/oppia/app/testing/ExplorationTestActivity.kt (hunk 10 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/testing/ExplorationTestActivity.kt]: Done

File: app/src/main/java/org/oppia/app/testing/ExplorationTestActivityPresenter.kt (hunk 11 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/testing/ExplorationTestActivityPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/topic/TopicActivity.kt (hunk 12 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/topic/TopicActivity.kt]: Done

File: app/src/main/java/org/oppia/app/topic/lessons/TopicLessonsFragmentPresenter.kt (hunk 13 above)
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/app/topic/lessons/TopicLessonsFragmentPresenter.kt]: Done

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt (hunk 14 above)
Comment (Sarthak2601) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Done

**Comment by Sarthak2601** [comment]
@rt4914 Done with all the nits. Self reviewed and removed unused imports and did one more nit as well. PTAL :)

**Review by rt4914 — APPROVED** [review]
@Sarthak2601 LGTM

**Comment by rt4914** [comment]
@Sarthak2601 Just resolve merge conflict and merge it. Thanks.

**Pull request merged**